packages/babel-parser/src/parser/expression.js

```diff
@@ -1158,7 +1171,11 @@ export default class ExpressionParser extends LValParser {
     const isPrivate = this.match(tt.hash);
 
     if (isPrivate) {
-      this.expectOnePlugin(["classPrivateProperties", "classPrivateMethods"]);
```

Comment:
I think throwing missing `privateIn` error for the common private property case will confuse users.
```js
class C {
  #p = 42;
}
```

Instead we can expect `privateIn` when we see `tt._in` after a private name.

Reply:
I don't understand the suggestion. I need to expect one of the 3 plugins here, or else I won't be able to parse the private name.

Reply:
I just realize that this line is never failed due to https://github.com/babel/babel/pull/11478/files#diff-80da704a8d2fc8d339ab57f154fb225eL429, which I address in #11478.

Base on #11478 and current implementation, if users are enabling `classPrivateProperties` but without `privateIn`, the following case does not throw missing `privateIn` error. ([REPL](https://babeljs.io/repl/build/21259/#?browsers=node%206&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEDC0G8BQ1oGIAO0C80CMA3CtALYAUAlItAE4CmALgK40B26WAluwwBacwAvkkFA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=script&lineWrap=false&presets=env%2Cstage-3%2Cenv&prettier=true&targets=Node-13&version=7.9.0%2Bpr.11372&externalPlugins=))
```js
class C {
  #p = 42;
  m() { return #p in this }
}
```
because we never really check `privateIn` as long as `classPrivateProperties` is enabled.

Can you add a test case to `packages/babel-parser/test/fixtures/experimental/_no-plugin`?

Reply:
We check at https://github.com/babel/babel/pull/11372/files#diff-56fea09eea2de51b8291db400d2f69beR1142. Added a test, we get `SyntaxError: Unexpected token` at the `#` in `#foo in {}`.

Reply:
Can we throw `the privateIn plugin is expected` when we see `#foo in {}` but `privateIn` is missing? Just like we [did](https://babeljs.io/repl/build/21296/#?browsers=node%2013&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEDC0G8BQ1oGIAO0C80CMA3EgL5A&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ctypescript%2Cenv&prettier=true&targets=&version=7.9.0%2Bpr.11478&externalPlugins=) for `classPrivateProperties`.

Reply:
There are several conflicting desires here:

1. Parsing smart pipelines, and giving good errors
2. Parsing invalid `#foo` expressions (not followed by `in`)
3. Parsing `#foo in {}` expressions, but not having plugin enabled

I don't have a good solution to all three of these. If you can push a commit, that'd be awesome.

Reply:
Pushed a commit that should address 2) and 3).

For 1) we are printing good errors when `|>` is seen but no `pipeline` plugin. I think current situation is acceptable and since they are competing proposals we can give better errors after any of them advances.